File: tree.py
# Implement a Tree data structure class

class Node():
    def __init__(self, data, children=None):
        self.data = data
        self.children = children or []

    # def __repr__(self):
        # implement a repr method for testing

    # Add child
    def add_child(self, node):
        self.children.append(node)

    # print children
    def print_children(self):
        for child in self.children:
            print(f'Child of {self.data}: ', child.data)

    def print_descendants(self):
        # create list for nodes to check instantiate it with current nodes children
        nodes_to_check = self.children.copy()
        # while nodes are in list
        while nodes_to_check:  # [] == false
            # pop off front of list to print value
            curr = nodes_to_check.pop(0)
            # print value
            print('Child of Parent 1', curr.data)
            # check if node has any children
            if curr.children:  # [] == false
                # add children to front of nodes to check list
                for child in curr.children:
                    # insert each child at front of list
                    nodes_to_check.insert(0, child)

    def is_child(self, data):
        nodes_to_check = self.children.copy()

        while nodes_to_check:
            curr = nodes_to_check.pop()
            if curr.data == data:
                print('True')
                return
            else:
                if curr.children:
                    for child in curr.children:
                        nodes_to_check.append(child)
        print('False')


node1 = Node(1)
node9 = Node(9)
node2 = Node(2)
node3 = Node(3)
node4 = Node(4)
node5 = Node(5)
node6 = Node(6)
node7 = Node(7)
node8 = Node(8)
node1.add_child(node9)
node1.add_child(node2)
node2.add_child(node4)
node4.add_child(node6)
node4.add_child(node5)
node6.add_child(node7)
node3.add_child(node8)
node1.add_child(node3)


node1.print_descendants()
node3.is_child(10)  # True

File: tree.js
/* ==== Build a Tree class ==== */

class Tree {

    constructor(data) {
        this.data = data;
        this.children = []
    }

    addChild(node) {
        this.children.push(node);
    }

    printChildren() {
        for(const child of this.children) {
            console.log(child.data)
        }
    }

    printDescendants(child) {
        if(this.children.length === 0) console.log('No descendants');
        if(child) {
            console.log(child.data);
            if(child.children) {
                for(const des of child.children) this.printDescendants(des);
            }
            return;
        }
        for(const child of this.children) {
           this.printDescendants(child)
        }
    }

    isChild(node) {
        for(const child of this.children) {
            if(child === node) return true;
        }
        return false;
    }

    // isDescendant(node) {
    //     for(const child of this.children) {
    //         if(child === node) return true
    //     }
    // }
}

// const tree1 = new Tree(1);
// const tree2 = new Tree(2);
// const tree3 = new Tree(3);
// const tree4 = new Tree(4);
// const tree5 = new Tree(5);

// tree1.addChild(tree2);
// tree1.addChild(tree3);
// tree2.addChild(tree4);
// tree2.addChild(tree5);

// // tree1.printChildren();
// tree1.printDescendants();
// console.log(tree2.isChild(tree5))

module.exports = Tree;